T/16 Introduce DataTransfer#files property, change the clipboardInput event and add the inputTranformation event

ma2ciek opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced `DataTransfer#files` property. Changed the `clipboardInput` event and added the `inputTranformation` event. Closes #16.

ma2ciek commented:
TODO: docs

ma2ciek commented:
@Reinmar Not sure whether this should be `clipboard` or `clipboard/clipboard`: https://github.com/ckeditor/ckeditor5-clipboard/pull/17/files#diff-2db77e7b10eb10844f23dd960369d74aL98

ma2ciek commented on src/clipboard.js:
```diff
@@ -95,7 +95,7 @@ export default class Clipboard extends Plugin {
 	 * @inheritDoc
 	 */
 	static get pluginName() {
-		return 'clipboard/clipboard';
```
@Reinmar Not sure whether this should be clipboard or clipboard/clipboard

Reinmar replied:
See documentation of `Plugin.pluginName`. It should be `clipboard/clipboard` – one is the package name (meaningful part of it) and the other is the module inside it.

pjasiun commented:
Note that this PR also introduced "files" in the dataTransfer object. It should be mentioned in the commit message.

pjasiun commented:
Also the `files` property miss documentation.

pjasiun commented on src/clipboard.js:
```diff
@@ -26,21 +26,35 @@ import HtmlDataProcessor from '@ckeditor/ckeditor5-engine/src/dataprocessor/html
- * 2. prevent the default action of the native `paste` event,
- * 3. fire {@link module:engine/view/document~Document#event:clipboardInput} with the clipboard data parsed to
+ * 2. prevent the default action of the native `paste` or `drop` event,
+ * 3. fire {@link module:engine/view/document~Document#event:input} with a
```
`input` event is not fired on the `Document` anymore, but on the `ClipboadObserver` now.

pjasiun commented on src/clipboard.js:
```diff
@@ -179,34 +191,28 @@ export default class Clipboard extends Plugin {
- * Fired with a content which comes from the clipboard (was pasted or dropped) and
+ * Fired with a `dataTransfer`, which comes from the clipboard (was pasted or dropped) and
  * should be processed in order to be inserted into the editor.
  * It's part of the {@link module:clipboard/clipboard~Clipboard "clipboard pipeline"}.
```
This is observers event now. Docs should be moved to the clipboard observer.

ma2ciek commented:
> Note that this PR also introduced "files" in the dataTransfer object. It should be mentioned in the commit message.

Fixed.

Added some docs improvements.

Reinmar commented:
You know guys that the `#files` prop makes for a perfect separate PR? ;) But ok... I'll review both things.

Reinmar commented on src/clipboardobserver.js:
```diff
@@ -22,17 +22,53 @@ export default class ClipboardObserver extends DomEventObserver {
+			// Prevent page refreshing.
+			data.preventDefault();
+
+			doc.fire( 'input', { dataTransfer: data.dataTransfer } );
```
This event is really confusing. There's a native`input` event. So this one would need to be called `clipboardInput`. Is it ok?

pjasiun replied:
That's fine for me.

Reinmar commented on src/clipboardobserver.js:
```diff
@@ -22,17 +22,53 @@ export default class ClipboardObserver extends DomEventObserver {
+		doc.on( 'drop', _handleInput, { priority: 'low' } );
+
+		function _handleInput( evt, data ) {
+			// Prevent page refreshing.
```
This is not true. This line doesn't prevent page refreshing. It prevents the default action (which only in case of file dropping can be page reload). No reason to comment it.

Reinmar commented on src/clipboardobserver.js:
```diff
@@ -22,17 +22,53 @@ export default class ClipboardObserver extends DomEventObserver {
+		doc.on( 'paste', _handleInput, { priority: 'low' } );
+		doc.on( 'drop', _handleInput, { priority: 'low' } );
+
+		function _handleInput( evt, data ) {
```
No reason to prefix this function name with `_`.

Reinmar commented on src/clipboardobserver.js:
```diff
@@ -22,17 +22,53 @@ export default class ClipboardObserver extends DomEventObserver {
-		this.domEventType = [ 'paste', 'copy', 'cut' ];
+		this.domEventType = [ 'paste', 'copy', 'cut', 'drop' ];
+
+		doc.on( 'paste', _handleInput, { priority: 'low' } );
```
This is unsafe and incorrect. `this.listenTo()` should be used. Those listeners should be gone together with the observer.

Reinmar commented on src/datatransfer.js:
```diff
@@ -42,4 +50,26 @@ export default class DataTransfer {
 		this._native.setData( type, data );
 	}
+
+	*_getFiles() {
```
Why is this function a generator if above we do `Array.from()` anyway? It doesn't make sense.

Reinmar replied:
Besides, it needs at least minimal docs. Or even better – move it out of the prototype as it doesn't make sense here (unless for testing purposes, but then document this too).

pjasiun replied:
That's because of the historical reasons. There were no `files` at the very beginning because there are some performance problems with reading files handlers, but then I realized that you need to get files handlers anyway to check if there are files, so the property was created. That's true that there is not a reason to keep it as a generator.

Reinmar commented on src/datatransfer.js:
```diff
@@ -42,4 +50,26 @@ export default class DataTransfer {
+		if ( files.length ) {
+			yield* files;
+		}
+		// // Chrome have empty DataTransfer.files, but let get files through the items interface.
```
Doubled `//`.

Reinmar commented on src/datatransfer.js:
```diff
@@ -42,4 +50,26 @@ export default class DataTransfer {
+		}
+	}
+
+	get types() {
```
Properties should be before methods and public stuff must be before protected and private stuff. Plus, it misses docs.

Reinmar commented on src/datatransfer.js:
```diff
@@ -19,6 +19,14 @@ export default class DataTransfer {
+		/**
+		 * The array of files created from the native `DataTransfer#files` and `DataTransfer#items` objects.
+		 *
+		 * @public
```
It's also `@readonly`.

Reinmar commented on src/datatransfer.js:
```diff
@@ -19,6 +19,14 @@ export default class DataTransfer {
 		this._native = nativeDataTransfer;
+
+		/**
+		 * The array of files created from the native `DataTransfer#files` and `DataTransfer#items` objects.
```
This is confusing – it indicates that `#files` is a combination of these two native properties while it's either one or another.

Reinmar commented on src/clipboard.js:
```diff
@@ -26,21 +26,35 @@ import HtmlDataProcessor from '@ckeditor/ckeditor5-engine/src/dataprocessor/html
+ *		this.listenTo( editor.editing.view, 'input', ( evt, data ) => {
+ *			const content = customTransform( data.dataTransfer.get( 'text/html' ) );
+ *			const transformedContent = transform( content );
+ * 			data.dataTransfer.set( 'text/html', transformedContent );
```
Indentation.

Reinmar commented on src/clipboard.js:
```diff
@@ -50,7 +64,7 @@ import HtmlDataProcessor from '@ckeditor/ckeditor5-engine/src/dataprocessor/html
  * a pasted text into a link can be implemented in this way:
  *
- *		this.listenTo( editor.editing.view, 'clipboardInput', ( evt, data ) => {
+ *		this.listenTo( editor.plugins.get( 'clipboard/clipboard' ), 'inputTransformation', ( evt, data ) => {
```
Hm... I'm not sure that it makes sense to showcase this method of retrieving a plugin. It may not be loaded and such code would break. OTOH, showing the right way would make this code unnecessarily long, so perhaps we can leave this.

Reinmar commented:
R- for now. I haven't checked all but the rest should be cosmetics.

ma2ciek commented:
It's ready for the review once again.
